File: Python/maze_utils/serializer.py
"""
Maze Serializers - 迷宫序列化工具
==================================

提供迷宫的序列化和反序列化功能：
- JSON格式
- 字典格式
- 文件读写
"""

import json
from typing import Dict, Any, List, Optional
from .maze import Maze, Cell, Direction


def to_dict(maze: Maze) -> Dict[str, Any]:
    """
    将迷宫转换为字典格式
    
    Args:
        maze: 迷宫对象
        
    Returns:
        字典表示
    """
    cells = []
    for y in range(maze.height):
        row = []
        for x in range(maze.width):
            cell = maze.get_cell(x, y)
            row.append({
                'walls': [d.name for d in cell.walls],
                'visited': cell.visited,
            })
        cells.append(row)
    
    return {
        'version': '1.0',
        'width': maze.width,
        'height': maze.height,
        'start': list(maze._start) if maze._start else [0, 0],
        'end': list(maze._end) if maze._end else [maze.width - 1, maze.height - 1],
        'cells': cells,
    }


def from_dict(data: Dict[str, Any]) -> Maze:
    """
    从字典创建迷宫
    
    Args:
        data: 字典数据
        
    Returns:
        迷宫对象
    """
    width = data['width']
    height = data['height']
    
    maze = Maze(width, height)
    
    # 设置起点和终点
    if 'start' in data:
        maze.start = tuple(data['start'])
    if 'end' in data:
        maze.end = tuple(data['end'])
    
    # 设置单元格
    cells_data = data.get('cells', [])
    for y, row_data in enumerate(cells_data):
        for x, cell_data in enumerate(row_data):
            cell = maze.get_cell(x, y)
            cell.walls = {Direction[d] for d in cell_data.get('walls', [])}
            cell.visited = cell_data.get('visited', False)
    
    return maze


def to_json(maze: Maze, indent: Optional[int] = 2) -> str:
    """
    将迷宫序列化为JSON字符串
    
    Args:
        maze: 迷宫对象
        indent: 缩进级别
        
    Returns:
        JSON字符串
    """
    return json.dumps(to_dict(maze), indent=indent, ensure_ascii=False)


def from_json(json_str: str) -> Maze:
    """
    从JSON字符串创建迷宫
    
    Args:
        json_str: JSON字符串
        
    Returns:
        迷宫对象
    """
    data = json.loads(json_str)
    return from_dict(data)


def save_to_file(maze: Maze, filepath: str) -> None:
    """
    将迷宫保存到文件
    
    Args:
        maze: 迷宫对象
        filepath: 文件路径
    """
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(to_json(maze))


def load_from_file(filepath: str) -> Maze:
    """
    从文件加载迷宫
    
    Args:
        filepath: 文件路径
        
    Returns:
        迷宫对象
    """
    with open(filepath, 'r', encoding='utf-8') as f:
        return from_json(f.read())


def to_binary(maze: Maze) -> bytes:
    """
    将迷宫转换为紧凑的二进制格式
    
    格式：
    - 2字节: 宽度 (uint16, big-endian)
    - 2字节: 高度 (uint16, big-endian)
    - 1字节: 起点X
    - 1字节: 起点Y
    - 1字节: 终点X
    - 1字节: 终点Y
    - N字节: 墙壁位图 (每个单元格4位，表示4个方向)
    
    Args:
        maze: 迷宫对象
        
    Returns:
        二进制数据
    """
    data = bytearray()
    
    # 尺寸
    data.extend(maze.width.to_bytes(2, 'big'))
    data.extend(maze.height.to_bytes(2, 'big'))
    
    # 起点和终点
    data.append(maze.start[0])
    data.append(maze.start[1])
    data.append(maze.end[0])
    data.append(maze.end[1])
    
    # 墙壁位图
    for y in range(maze.height):
        for x in range(0, maze.width, 2):
            cell1 = maze.get_cell(x, y)
            # 每个单元格用4位表示4个墙
            walls1 = 0
            if Direction.NORTH in cell1.walls:
                walls1 |= 0x08
            if Direction.SOUTH in cell1.walls:
                walls1 |= 0x04
            if Direction.EAST in cell1.walls:
                walls1 |= 0x02
            if Direction.WEST in cell1.walls:
                walls1 |= 0x01
            
            if x + 1 < maze.width:
                cell2 = maze.get_cell(x + 1, y)
                walls2 = 0
                if Direction.NORTH in cell2.walls:
                    walls2 |= 0x08
                if Direction.SOUTH in cell2.walls:
                    walls2 |= 0x04
                if Direction.EAST in cell2.walls:
                    walls2 |= 0x02
                if Direction.WEST in cell2.walls:
                    walls2 |= 0x01
                data.append((walls1 << 4) | walls2)
            else:
                data.append(walls1 << 4)
    
    return bytes(data)


def from_binary(data: bytes) -> Maze:
    """
    从二进制数据创建迷宫
    
    Args:
        data: 二进制数据
        
    Returns:
        迷宫对象
    """
    if len(data) < 8:
        raise ValueError("Invalid binary maze data: too short")
    
    # 读取尺寸
    width = int.from_bytes(data[0:2], 'big')
    height = int.from_bytes(data[2:4], 'big')
    
    maze = Maze(width, height)
    
    # 读取起点和终点
    maze.start = (data[4], data[5])
    maze.end = (data[6], data[7])
    
    # 读取墙壁位图
    offset = 8
    for y in range(height):
        for x in range(0, width, 2):
            byte = data[offset]
            offset += 1
            
            # 第一个单元格
            cell1 = maze.get_cell(x, y)
            walls1 = (byte >> 4) & 0x0F
            cell1.walls = set()
            if walls1 & 0x08:
                cell1.walls.add(Direction.NORTH)
            if walls1 & 0x04:
                cell1.walls.add(Direction.SOUTH)
            if walls1 & 0x02:
                cell1.walls.add(Direction.EAST)
            if walls1 & 0x01:
                cell1.walls.add(Direction.WEST)
            
            # 第二个单元格
            if x + 1 < width:
                cell2 = maze.get_cell(x + 1, y)
                walls2 = byte & 0x0F
                cell2.walls = set()
                if walls2 & 0x08:
                    cell2.walls.add(Direction.NORTH)
                if walls2 & 0x04:
                    cell2.walls.add(Direction.SOUTH)
                if walls2 & 0x02:
                    cell2.walls.add(Direction.EAST)
                if walls2 & 0x01:
                    cell2.walls.add(Direction.WEST)
    
    return maze


def to_csv(maze: Maze) -> str:
    """
    将迷宫导出为CSV格式（用于数据分析）
    
    格式：每行一个单元格，包含坐标和墙壁信息
    
    Args:
        maze: 迷宫对象
        
    Returns:
        CSV字符串
    """
    lines = ['x,y,north_wall,south_wall,east_wall,west_wall']
    
    for y in range(maze.height):
        for x in range(maze.width):
            cell = maze.get_cell(x, y)
            lines.append(','.join([
                str(x),
                str(y),
                '1' if Direction.NORTH in cell.walls else '0',
                '1' if Direction.SOUTH in cell.walls else '0',
                '1' if Direction.EAST in cell.walls else '0',
                '1' if Direction.WEST in cell.walls else '0',
            ]))
    
    return '\n'.join(lines)

File: Go/maze_utils/serializer.go
package maze_utils

import (
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// MazeJSON JSON 序列化格式
type MazeJSON struct {
	Width  int            `json:"width"`
	Height int            `json:"height"`
	Grid   [][]CellJSON   `json:"grid"`
	Start  [2]int         `json:"start"`
	End    [2]int         `json:"end"`
	Seed   int64          `json:"seed"`
}

// CellJSON 单元格的 JSON 格式
type CellJSON struct {
	X     int  `json:"x"`
	Y     int  `json:"y"`
	Walls [4]bool `json:"walls"` // N, E, S, W
}

// ToJSON 将迷宫序列化为 JSON 字符串
func ToJSON(m *Maze) (string, error) {
	data := MazeJSON{
		Width:  m.Width,
		Height: m.Height,
		Start:  m.Start,
		End:    m.End,
		Seed:   m.Seed,
		Grid:   make([][]CellJSON, m.Height),
	}

	for y := 0; y < m.Height; y++ {
		data.Grid[y] = make([]CellJSON, m.Width)
		for x := 0; x < m.Width; x++ {
			cell := m.Grid[y][x]
			data.Grid[y][x] = CellJSON{
				X:     cell.X,
				Y:     cell.Y,
				Walls: cell.Walls,
			}
		}
	}

	bytes, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return "", err
	}
	return string(bytes), nil
}

// FromJSON 从 JSON 字符串反序列化迷宫
func FromJSON(jsonStr string) (*Maze, error) {
	var data MazeJSON
	err := json.Unmarshal([]byte(jsonStr), &data)
	if err != nil {
		return nil, err
	}

	m := NewMaze(data.Width, data.Height)
	m.Start = data.Start
	m.End = data.End
	m.Seed = data.Seed

	for y := 0; y < data.Height; y++ {
		for x := 0; x < data.Width; x++ {
			m.Grid[y][x].Walls = data.Grid[y][x].Walls
		}
	}

	return m, nil
}

// ToMap 将迷宫转换为 map 格式
func ToMap(m *Maze) map[string]interface{} {
	grid := make([][]map[string]interface{}, m.Height)
	for y := 0; y < m.Height; y++ {
		grid[y] = make([]map[string]interface{}, m.Width)
		for x := 0; x < m.Width; x++ {
			cell := m.Grid[y][x]
			grid[y][x] = map[string]interface{}{
				"x":     cell.X,
				"y":     cell.Y,
				"walls": []bool{cell.Walls[0], cell.Walls[1], cell.Walls[2], cell.Walls[3]},
			}
		}
	}

	return map[string]interface{}{
		"width":  m.Width,
		"height": m.Height,
		"grid":   grid,
		"start":  []int{m.Start[0], m.Start[1]},
		"end":    []int{m.End[0], m.End[1]},
		"seed":   m.Seed,
	}
}

// FromMap 从 map 格式创建迷宫
func FromMap(data map[string]interface{}) (*Maze, error) {
	width := int(data["width"].(float64))
	height := int(data["height"].(float64))

	m := NewMaze(width, height)

	if start, ok := data["start"].([]interface{}); ok {
		m.Start = [2]int{int(start[0].(float64)), int(start[1].(float64))}
	}
	if end, ok := data["end"].([]interface{}); ok {
		m.End = [2]int{int(end[0].(float64)), int(end[1].(float64))}
	}
	if seed, ok := data["seed"].(float64); ok {
		m.Seed = int64(seed)
	}

	gridData, ok := data["grid"].([]interface{})
	if !ok {
		return nil, fmt.Errorf("invalid grid data")
	}

	for y, row := range gridData {
		rowData, ok := row.([]interface{})
		if !ok {
			return nil, fmt.Errorf("invalid row data at y=%d", y)
		}
		for x, cellData := range rowData {
			cell, ok := cellData.(map[string]interface{})
			if !ok {
				return nil, fmt.Errorf("invalid cell data at (%d,%d)", x, y)
			}
			wallsData, ok := cell["walls"].([]interface{})
			if ok {
				for i, w := range wallsData {
					if i < 4 {
						m.Grid[y][x].Walls[i] = w.(bool)
					}
				}
			}
		}
	}

	return m, nil
}

// ToBinary 将迷宫序列化为二进制格式
// 格式: [width:2][height:2][start_x:2][start_y:2][end_x:2][end_y:2][seed:8]
//       [walls: 每个单元格1字节(NESW低4位)]
func ToBinary(m *Maze) []byte {
	// 计算大小: 2 + 2 + 2 + 2 + 2 + 2 + 8 + (width * height) = 20 + (width * height)
	size := 20 + m.Width*m.Height
	data := make([]byte, size)
	pos := 0

	// 写入宽高
	data[pos] = byte(m.Width >> 8)
	data[pos+1] = byte(m.Width)
	pos += 2
	data[pos] = byte(m.Height >> 8)
	data[pos+1] = byte(m.Height)
	pos += 2

	// 写入起点终点
	data[pos] = byte(m.Start[0] >> 8)
	data[pos+1] = byte(m.Start[0])
	pos += 2
	data[pos] = byte(m.Start[1] >> 8)
	data[pos+1] = byte(m.Start[1])
	pos += 2
	data[pos] = byte(m.End[0] >> 8)
	data[pos+1] = byte(m.End[0])
	pos += 2
	data[pos] = byte(m.End[1] >> 8)
	data[pos+1] = byte(m.End[1])
	pos += 2

	// 写入种子
	for i := 7; i >= 0; i-- {
		data[pos+i] = byte(m.Seed >> (8 * (7 - i)))
	}
	pos += 8

	// 写入墙壁数据
	for y := 0; y < m.Height; y++ {
		for x := 0; x < m.Width; x++ {
			cell := m.Grid[y][x]
			var b byte
			if cell.Walls[North] {
				b |= 1 << 3
			}
			if cell.Walls[East] {
				b |= 1 << 2
			}
			if cell.Walls[South] {
				b |= 1 << 1
			}
			if cell.Walls[West] {
				b |= 1 << 0
			}
			data[pos] = b
			pos++
		}
	}

	return data
}

// FromBinary 从二进制格式反序列化迷宫
func FromBinary(data []byte) (*Maze, error) {
	if len(data) < 18 {
		return nil, fmt.Errorf("invalid binary data: too short")
	}

	pos := 0

	// 读取宽高
	width := int(data[pos])<<8 | int(data[pos+1])
	pos += 2
	height := int(data[pos])<<8 | int(data[pos+1])
	pos += 2

	if width <= 0 || height <= 0 || width > 10000 || height > 10000 {
		return nil, fmt.Errorf("invalid maze dimensions: %dx%d", width, height)
	}

	// 读取起点终点
	startX := int(data[pos])<<8 | int(data[pos+1])
	pos += 2
	startY := int(data[pos])<<8 | int(data[pos+1])
	pos += 2
	endX := int(data[pos])<<8 | int(data[pos+1])
	pos += 2
	endY := int(data[pos])<<8 | int(data[pos+1])
	pos += 2

	// 读取种子
	var seed int64
	for i := 0; i < 8; i++ {
		seed = seed<<8 | int64(data[pos])
		pos++
	}

	// 验证数据长度
	expectedLen := 20 + width*height
	if len(data) < expectedLen {
		return nil, fmt.Errorf("invalid binary data: expected %d bytes, got %d", expectedLen, len(data))
	}

	// 创建迷宫
	m := NewMaze(width, height)
	m.Start = [2]int{startX, startY}
	m.End = [2]int{endX, endY}
	m.Seed = seed

	// 读取墙壁数据
	for y := 0; y < height; y++ {
		for x := 0; x < width; x++ {
			b := data[pos]
			pos++
			m.Grid[y][x].Walls[North] = (b & (1 << 3)) != 0
			m.Grid[y][x].Walls[East] = (b & (1 << 2)) != 0
			m.Grid[y][x].Walls[South] = (b & (1 << 1)) != 0
			m.Grid[y][x].Walls[West] = (b & (1 << 0)) != 0
		}
	}

	return m, nil
}

// ToCSV 将迷宫导出为 CSV 格式
// 每行: x,y,north,east,south,west
func ToCSV(m *Maze) string {
	var sb strings.Builder
	sb.WriteString("x,y,north,east,south,west\n")

	for y := 0; y < m.Height; y++ {
		for x := 0; x < m.Width; x++ {
			cell := m.Grid[y][x]
			sb.WriteString(fmt.Sprintf("%d,%d,%t,%t,%t,%t\n",
				x, y,
				cell.Walls[North],
				cell.Walls[East],
				cell.Walls[South],
				cell.Walls[West],
			))
		}
	}

	return sb.String()
}

// FromCSV 从 CSV 格式导入迷宫
func FromCSV(csv string) (*Maze, error) {
	lines := strings.Split(csv, "\n")
	if len(lines) < 2 {
		return nil, fmt.Errorf("invalid CSV: no data")
	}

	// 解析所有单元格
	cellMap := make(map[[2]int][4]bool)
	var maxX, maxY int

	for i, line := range lines[1:] {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		parts := strings.Split(line, ",")
		if len(parts) != 6 {
			return nil, fmt.Errorf("invalid CSV line %d: expected 6 fields", i+2)
		}

		x, _ := strconv.Atoi(parts[0])
		y, _ := strconv.Atoi(parts[1])

		if x > maxX {
			maxX = x
		}
		if y > maxY {
			maxY = y
		}

		walls := [4]bool{}
		walls[North] = parts[2] == "true"
		walls[East] = parts[3] == "true"
		walls[South] = parts[4] == "true"
		walls[West] = parts[5] == "true"

		cellMap[[2]int{x, y}] = walls
	}

	// 创建迷宫
	m := NewMaze(maxX+1, maxY+1)
	for pos, walls := range cellMap {
		if pos[0] >= 0 && pos[0] < m.Width && pos[1] >= 0 && pos[1] < m.Height {
			m.Grid[pos[1]][pos[0]].Walls = walls
		}
	}

	return m, nil
}

// SaveToFile 将迷宫保存到文件
func SaveToFile(m *Maze, filename string) error {
	data, err := ToJSON(m)
	if err != nil {
		return err
	}
	return os.WriteFile(filename, []byte(data), 0644)
}

// LoadFromFile 从文件加载迷宫
func LoadFromFile(filename string) (*Maze, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	return FromJSON(string(data))
}

// PathToJSON 将路径序列化为 JSON
func PathToJSON(path Path) string {
	data := make([][2]int, len(path))
	copy(data, path)
	bytes, _ := json.Marshal(data)
	return string(bytes)
}

// PathFromJSON 从 JSON 反序列化路径
func PathFromJSON(jsonStr string) (Path, error) {
	var path Path
	err := json.Unmarshal([]byte(jsonStr), &path)
	return path, err
}